### ISSUE TITLE
current status of n_jobs in v2.10?

### ISSUE THREAD

**Issue by rbelew** [optuna/optuna#3134]
### Question

I find a merge Jan 29 using `concurrent.futures` https://github.com/optuna/optuna/pull/2269, then 
I find  the `njobs` parameter was deprecated here https://github.com/optuna/optuna/pull/2560 on Apr 1.  And now there is discussion about reintroducing it here https://github.com/optuna/optuna/issues/2937 on Sept 18.  I also found an older work-arounds, using `joblib` https://gist.github.com/sile/2c329d0cb38f35e25a8c7bab44edbb63. 

I am wondering what the current recommended procedure for multi-core environments using v2.10.0 optuna.  What does the implementation currently do with `n_jobs` specified?  It seems to produce odd, non-montonic best values (see example below).

### Environment
```
>>> print(op.__version__) 
2.10.0
>>> print(platform.python_version()) 
3.8.5
>>> print(platform.platform()) 
macOS-10.15.7-x86_64-i386-64bit
```

### Error messages, stack traces, or logs

```
Trial 101 finished with best value: 3.051867903393622 and parameters: {'R0': 1.0712727856991768, 'alpha': 0.9121462925283631, 'gamma': 1.1823702204733835, 'mu': 0.036417631166185516}. 
Trial 188 finished with best value: 2.2365535937364998 and parameters: {'R0': 1.0694600337354254, 'alpha': 0.8999322001243446, 'gamma': 1.3864523757139084, 'mu': 0.02695870506276768}. 
Trial 277 finished with best value: 2.088390682306482 and parameters: {'R0': 1.054260021970593, 'alpha': 1.5693093881559335, 'gamma': 1.469040063506123, 'mu': 0.03048413180433311}. 
Trial 276 finished with best value: 15.34937870021244 and parameters: {'R0': 1.1377582001694242, 'alpha': 0.6932468664055964, 'gamma': 1.236927077386912, 'mu': 0.04018656386490023}. 
Trial 292 finished with best value: 1.3179174767632489 and parameters: {'R0': 1.0631375246172654, 'alpha': 1.5265556047812037, 'gamma': 1.3094584471794986, 'mu': 0.025065419768277475}. 
Trial 293 finished with best value: 1.6933340465788846 and parameters: {'R0': 1.0665697529499913, 'alpha': 1.5104281490808449, 'gamma': 1.4077384827255122, 'mu': 0.02765503785586052}. 

```
### Steps to reproduce

N/A

### Reproducible examples

N/A

**Comment by rbelew** [comment]
hello @keisuke-umezawa thanks very much for your reply!  
The lines i provided last time were using the [logging callback you suggested ](https://optuna.readthedocs.io/en/stable/faq.html?highlight=logging#how-can-i-output-a-log-only-when-the-best-value-is-updated).

i've done another experiment to confirm, with n_jobs=4, and then did a simple query against your database:
```
SELECT trials.trial_id,trials.datetime_complete,trial_values.value   from trials JOIN trial_values on
  trials.trial_id=trial_values.trial_id
```
here are the 10 best values found:

```
trial_id	timeComplete	value
615	15:32:09	1.20842160689863
926	15:32:41	1.29022892788981
523	15:32:01	1.29713495757614
866	15:32:35	1.45750623757587
538	15:32:02	1.4753850876484
537	15:32:02	1.50236931729865
614	15:32:09	1.57503937942763
826	15:32:31	1.58075911148432
883	15:32:36	1.5823593890732
569	15:32:05	1.60183105975086
```
trial#615 found a better value than the later trial#926, trial#523 better than trial#866, etc.  i'm attaching the complete query output.  Sorting by trial_id shows the batches of 4 nearly synchronous launches.

i can also send your complete database if useful.

extracting a minimal example from my code will be an effort, but i can try to do that if you think it's worthwhile.

Thanks again,  Rik

[trials.csv](https://github.com/optuna/optuna/files/7637795/trials.csv)

**Comment by rbelew** [comment]
i just tried the joblib workaround, and it also shows the same non-monotonic ordering of trial values over time?

**Comment by himkt** [comment]
Thank you @rbelew for the explanation. I reproduced the results by the following code (@hvy wrote it):

```python
import optuna
import random
import time


# Turn off optuna log notes.
optuna.logging.set_verbosity(optuna.logging.WARN)


def objective(trial):
    x = trial.suggest_float("x", 0, 1)
    return x ** 2


def logging_callback(study, frozen_trial):
    previous_best_value = study.user_attrs.get("previous_best_value", None)

    if previous_best_value != study.best_value:
        study.set_user_attr("previous_best_value", study.best_value)

        print(
            "Trial {} finished with best value: {} and parameters: {}. ".format(
            frozen_trial.number,
            frozen_trial.value,
            frozen_trial.params,
            )
        )


study = optuna.create_study()
study.optimize(objective, n_trials=100, callbacks=[logging_callback], n_jobs=10)
```

And I found this code doesn't work as expected in multi-{thread,process} situations.

- In multi-thread situation, some threads may update a system attribute simultaneously (race-condition)
- In multi-process situation, I'm not sure but I think each process doesn't share the study (which causes the discrepancy and odd behavior)
  - https://gist.github.com/sile/2c329d0cb38f35e25a8c7bab44edbb63#file-multi_process_storage-py-L59-L62

In summary, reporting best value occurance is difficult with multi-{thread,process} optimization.
We have to update our documentation for letting users know this limitation.

@hvy Please add a comment if I misunderstand something or you have supplement for this topic.

**Comment by rbelew** [comment]
Thanks to you both @hvy and @himkt for your quick attention to this.

perhaps we are saying the same thing, but my concern is not just with "reporting best value occurance" but rather whether optuna is making ANY use of search on separate threads?  the `trials` database i reported makes me think it is not?

i've not experimented with the multi-thread, extra-Pythonic solutions (ala your  [tutorial](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/004_distributed.html) so i can't speak to those.  But you have found that they effectively make use of the parallel search results?

thanks again for your good works,  Rik

**Comment by himkt** [comment]
@rbelew 
Sorry for my poor explanation. Let me try to answer your concerns again.

> whether optuna is making ANY effective use of the search done on separate threads?

Optuna makes effective use of the search done on separate threads.

Optuna can retrieve completed **trials** from a storage regardless of where the search is done.
`trial` is created for each thread or process therefore no racing issue happen.
On the other hand, system attribute is a global thing for the study,
which is updated simultaneously by multiple threads or processes (causing racing issue).

> trials database i reported makes me think it is not?

I checked your entire results using visualization module in Optuna.
It looks that a sampler tried a wide range of parameters during the early trials (<100)
to find a plausible area and continued the search based on the observation in the later trials.

<img src="https://user-images.githubusercontent.com/5164000/144527720-26c6ff71-3134-4a3a-a989-d141a225a805.png" width="500">

As you pointed out, the best performance history shows non-monotonicity.
Sampler may sample a parameter that performs worse even after it finds a parameter
that performs best to explore the chance to further improve of performance so the result could be non-monotonic.
